test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java

```diff
@@ -1081,6 +1081,9 @@ private synchronized void reset(boolean wipeData) throws IOException {
             final Settings.Builder settings = Settings.builder();
             settings.put(Node.NODE_MASTER_SETTING.getKey(), true);
             settings.put(Node.NODE_DATA_SETTING.getKey(), false);
+            if (prevNodeCount == 0 && autoManageMinMasterNodes) {
+                settings.put(INITIAL_MASTER_NODE_COUNT_SETTING.getKey(), numSharedDedicatedMasterNodes + numSharedDataNodes);
```

Comment:
I think this needs to be 
```
settings.put(INITIAL_MASTER_NODE_COUNT_SETTING.getKey(), numSharedDedicatedMasterNodes +
                    (numSharedDedicatedMasterNodes > 0 ? 0 : numSharedDataNodes));
```

because of the condition further below where we make the shared data nodes data-only nodes when there are dedicated master nodes.